ddlitlab2024/ml/training/train.py

```diff
@@ -1,130 +1,129 @@
+        use_action_history=True,
+        num_action_history_encoder_layers=2,
+        max_action_context_length=100,
+        use_imu=True,
+        imu_orientation_embedding_method=IMUEncoder.OrientationEmbeddingMethod.QUATERNION,
+        num_imu_encoder_layers=2,
+        max_imu_context_length=100,
+        use_joint_states=True,
+        joint_state_encoder_layers=2,
+        max_joint_state_context_length=100,
+        use_images=True,
+        image_sequence_encoder_type=SequenceEncoderType.TRANSFORMER,
+        image_encoder_type=ImageEncoderType.RESNET18,
+        num_image_sequence_encoder_layers=1,
+        max_image_context_length=10,
+        num_decoder_layers=4,
+        trajectory_prediction_length=10,
     ).to(device)
 
     # Add normalization parameters to the model
-    model.mean = torch.tensor(means.values).to(device)
-    model.std = torch.tensor(stds.values).to(device)
+    model.mean = normalizer.mean
+    model.std = normalizer.std
 
+    # Utilize an Exponential Moving Average (EMA) for the model to smooth out the training process
     ema = EMA(model, beta=0.9999)
 
+    # Create optimizer and learning rate scheduler
     optimizer = torch.optim.AdamW(model.parameters(), lr=lr)
-    lr_scheduler = torch.optim.lr_scheduler.OneCycleLR(
-        optimizer, max_lr=lr, total_steps=epochs * (num_samples // batch_size)
-    )
+    lr_scheduler = torch.optim.lr_scheduler.OneCycleLR(optimizer, max_lr=lr, total_steps=epochs * len(dataloader))
 
+    # Create diffusion noise scheduler
     scheduler = DDIMScheduler(beta_schedule="squaredcos_cap_v2", clip_sample=False)
-    scheduler.config["num_train_timesteps"] = train_timesteps
+    scheduler.config["num_train_timesteps"] = train_denoising_timesteps
 
     # Training loop
     for epoch in range(epochs):  # Number of training epochs
         mean_loss = 0
-        # Shuffle the data for each epoch
-        real_trajectories = real_trajectories[torch.randperm(real_trajectories.size(0))]
 
-        for batch in tqdm(range(num_samples // batch_size)):
-            targets = real_trajectories[batch * batch_size : (batch + 1) * batch_size].to(device)
+        # Iterate over the dataset
+        for i, batch in enumerate(tqdm(dataloader)):
+            # Move the data to the device
+            batch = {k: v.to(device) for k, v in asdict(batch).items()}
+
+            # Extract the target actions
+            joint_targets = batch["joint_command"]
 
-            # Split the data into past actions and noisy action predictions
-            past_actions = targets[:, :action_context_length]
-            target_actions = targets[:, action_context_length:]
+            # Normalize the target actions
+            joint_targets = normalizer.normalize(joint_targets)
 
+            # Reset the gradients
             optimizer.zero_grad()
 
             # Sample a random timestep for each trajectory in the batch
             random_timesteps = (
```

Comment:
I'm confused as to what is random_timesteps here.
What dimensions does the tensor have? Is `joint_targets.size(0)` here 20, because of 20 degrees of freedom?

Additionally, why do we send the result of `randint` to device and not the whole `random_timesteps`?

Also we have a typing mismatch, because `random_timesteps` is not explicitly an `IntTensor`, which `add_noise` expects.